src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -9629,64 +9620,29 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
         }
 
         /// <summary>
-        /// Returns one of the methods from the method group if all methods in the method group
-        /// have the same signature, ignoring parameter names and custom modifiers. The particular
-        /// method returned is not important since the caller is interested in the signature only.
+        /// Returns one of the methods from the method group if all instance methods, or extension methods
+        /// in the nearest scope, have the same signature ignoring parameter names and custom modifiers.
+        /// The particular method returned is not important since the caller is interested in the signature only.
         /// </summary>
         private MethodSymbol? GetUniqueSignatureFromMethodGroup(BoundMethodGroup node)
         {
-            MethodSymbol? method = null;
-            foreach (var m in node.Methods)
-            {
-                switch (node.ReceiverOpt)
-                {
-                    case BoundTypeExpression:
-                    case null: // if `using static Class` is in effect, the receiver is missing
-                        if (!m.IsStatic) continue;
-                        break;
-                    case BoundThisReference { WasCompilerGenerated: true }:
-                        break;
-                    default:
-                        if (m.IsStatic) continue;
-                        break;
-                }
-                if (!isCandidateUnique(ref method, m))
-                {
-                    return null;
-                }
-            }
-            if (node.SearchExtensionMethods)
-            {
-                var receiver = node.ReceiverOpt!;
-                foreach (var scope in new ExtensionMethodScopes(this))
-                {
-                    var methodGroup = MethodGroup.GetInstance();
-                    PopulateExtensionMethodsFromSingleBinder(scope, methodGroup, node.Syntax, receiver, node.Name, node.TypeArgumentsOpt, BindingDiagnosticBag.Discarded);
-                    foreach (var m in methodGroup.Methods)
-                    {
-                        if (m.ReduceExtensionMethod(receiver.Type, Compilation) is { } reduced &&
-                            !isCandidateUnique(ref method, reduced))
-                        {
-                            methodGroup.Free();
-                            return null;
-                        }
-                    }
-                    methodGroup.Free();
-                }
-            }
+            var method = selectMethodForSignature(node);
+
             if (method is null)
             {
                 return null;
             }
-            int n = node.TypeArgumentsOpt.IsDefaultOrEmpty ? 0 : node.TypeArgumentsOpt.Length;
-            if (method.Arity != n)
+
+            int arity = node.TypeArgumentsOpt.IsDefaultOrEmpty ? 0 : node.TypeArgumentsOpt.Length;
+            if (method.Arity != arity)
```

Comment:
>if (method.Arity != arity) [](http://example.com/codeflow?start=12&length=26)

It feels like the arity check should be part of method selection. I.e. ```selectMethodForSignature``` should filter out methods that do not satisfy the target arity.

Reply:
Also, should the signature matching check be performed after type argument substitution?
Should candidates that violate constraints be filtered out?

Reply:
> It feels like the arity check should be part of method selection

There is an arity check as part of method selection already, both for instance and extension methods, but they allow zero arity member accesses to find methods of any arity. 
If this finds a method with non-zero arity, we must reject it as we can't infer the type arguments. This is illustrated by `MethodGroup_ScopeByScope_NoTypeArguments` amongst other tests.

I discussed adjusting that logic (zero arity matches any) for natural type determination with @cston but we identified scenarios where that would be undesirable:

```
using N;

var z = new C().M;
z();

public class C { }

public static class E1
{
    public static void M<T>(this T c) { } // we want to pick this method
}

namespace N
{
    public static class E2
    {
        public static void M(this C c) { }
    }
}
```

> Should candidates that violate constraints be filtered out?

That's a good question. I think that would make sense.
Let's confirm and handle as a separate issue since this change is focused on adopting a scope-by-scope approach.
Filed [https://github.com/dotnet/roslyn/issues/69222](https://github.com/dotnet/roslyn/issues/69222)

Reply:
I see that the same check was performed before, and was also performed at the end. So, there is no behavior change in this regard.